test/integration/navsat.cc

```diff
@@ -0,0 +1,332 @@
+  sensorNoise->SetAltitude(altitude);
+  EXPECT_DOUBLE_EQ(altitude, sensor->Altitude());
+  EXPECT_DOUBLE_EQ(altitude, sensorNoise->Altitude());
+
+  lat += IGN_DTOR(20);
+  lon += IGN_DTOR(20);
+  altitude += 100;
+  sensor->SetPosition(lat, lon, altitude);
+  sensorNoise->SetPosition(lat, lon, altitude);
+  EXPECT_EQ(lat, sensor->Latitude());
+  EXPECT_EQ(lat, sensorNoise->Latitude());
+  EXPECT_EQ(lon, sensor->Longitude());
+  EXPECT_EQ(lon, sensorNoise->Longitude());
+  EXPECT_DOUBLE_EQ(altitude, sensor->Altitude());
+  EXPECT_DOUBLE_EQ(altitude, sensorNoise->Altitude());
+
+  math::Vector3d velocity{1.0, 2.0, 3.0};
+  sensor->SetVelocity(velocity);
+  sensorNoise->SetVelocity(velocity);
+  EXPECT_EQ(velocity, sensor->Velocity());
+  EXPECT_EQ(velocity, sensorNoise->Velocity());
+
+  // verify msg received on the topic
+  WaitForMessageTestHelper<msgs::NavSat> msgHelper(topic);
+  sensor->Update(std::chrono::steady_clock::duration(1s));
+  EXPECT_TRUE(msgHelper.WaitForMessage()) << msgHelper;
+  auto msg = msgHelper.Message();
+  EXPECT_EQ(1, msg.header().stamp().sec());
+  EXPECT_EQ(0, msg.header().stamp().nsec());
+  EXPECT_DOUBLE_EQ(lat.Degree(), msg.latitude_deg());
+  EXPECT_DOUBLE_EQ(lon.Degree(), msg.longitude_deg());
+  EXPECT_DOUBLE_EQ(altitude, msg.altitude());
+  EXPECT_DOUBLE_EQ(velocity.X(), msg.velocity_east());
+  EXPECT_DOUBLE_EQ(velocity.Y(), msg.velocity_north());
+  EXPECT_DOUBLE_EQ(velocity.Z(), msg.velocity_up());
+
+  // verify msg with noise received on the topic
+  WaitForMessageTestHelper<msgs::NavSat> msgHelperNoise(topicNoise);
+  sensorNoise->Update(std::chrono::steady_clock::duration(1s));
+  EXPECT_TRUE(msgHelperNoise.WaitForMessage()) << msgHelperNoise;
+  auto msgNoise = msgHelperNoise.Message();
+  EXPECT_EQ(1, msgNoise.header().stamp().sec());
+  EXPECT_EQ(0, msgNoise.header().stamp().nsec());
+  EXPECT_FALSE(math::equal(lat.Degree(), msgNoise.latitude_deg()));
+  EXPECT_FALSE(math::equal(lon.Degree(), msgNoise.longitude_deg()));
+  EXPECT_FALSE(math::equal(altitude, msgNoise.altitude()));
+  EXPECT_FALSE(math::equal(velocity.X(), msgNoise.velocity_east()));
+  EXPECT_FALSE(math::equal(velocity.Y(), msgNoise.velocity_north()));
+  EXPECT_FALSE(math::equal(velocity.Z(), msgNoise.velocity_up()));
+}
+
+/////////////////////////////////////////////////
+TEST_F(NavSatSensorTest, Topic)
+{
+  const std::string name = "TestNavSat";
+  const double updateRate = 30;
+  const bool alwaysOn = 1;
+  const bool visualize = 1;
+  auto sensorPose = math::Pose3d();
+
+  // Factory
+  SensorFactory factory;
+
+  // Default topic
+  {
+    const std::string topic;
+    auto navsatSdf = NavSatToSdf(name, sensorPose, updateRate, topic,
+        alwaysOn, visualize);
+
+    auto navsat = factory.CreateSensor<NavSatSensor>(navsatSdf);
+    ASSERT_NE(nullptr, navsat);
+
+    EXPECT_EQ("/navsat", navsat->Topic());
+  }
+
+  // Convert to valid topic
+  {
+    const std::string topic = "/topic with spaces/@~characters//";
```

Comment:
where is the topic validity conversion taking place?

Reply:
It happens in `Sensor::SetTopic`, which is common to all sensors. If I remember correctly when I introduced topic validation in https://github.com/ignitionrobotics/ign-sensors/pull/33, the main goal of having these checks for every sensor was to make sure were correctly calling `SetTopic` for all of them :thinking: